### PR TITLE
[device] VecDequeue instead of ringbuffer

### PR DESCRIPTION
**Pull request by nickfarrow**

### CODE REVIEW

File: device/src/io.rs [hunk 1]
```diff
@@ -74,7 +74,7 @@ where
             return false;
         }
         let (progress, found) = frostsnap_comms::make_progress_on_magic_bytes::<D>(
-            self.ring_buffer.drain(),
+            self.ring_buffer.drain(..),
```
Comment (LLFourn) [inline device/src/io.rs:77]: I don't think this is correct:

[image](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.drain)

We need `drain` not to drop the entire buffer. We only want it to drop things from the buffer if they are iterated. It doesn't look like rust provides this built in which is weird. I think you can just do something like:

```rust
core::iter::from_fn(|| self.ring_buffer.pop_front())
```

**Review by LLFourn — CHANGES_REQUESTED** [review]
Need to look carefully at `drain`

File: device/src/io.rs (hunk 1 above)
Comment (nickfarrow) [reply device/src/io.rs]: Ah great spot, I didn't think to check if drain would behave differently.

So with this `drain`, if a device ingested magic bytes quickly followed by another message, the magic bytes would get read but the message in the buffer would be dropped.

This fix looks good!

**Comment by nickfarrow** [comment]
Took your fix for `drain`, also removed one of the drains by just using `.into::<Vec<u8>>()`

**Review by LLFourn — APPROVED** [review]
LGTM. @nickfarrow feel free to merge if you tested this a bit.

**Pull request merged**